Added UML class diagram to give overview of the most important classes

titusjan opened the pull request:
As discussed in issue #848, a UML diagram that gives an overview of the most important classes can really help with understanding the PyQtGraph library. 

I created a diagram in [PlantUML](https://plantuml.com/), as suggested. I therefore made this PR so that people have can access to the source file. I think it would be a useful addition to the documentation but I don't know where (and how) to add it. For now I just added it to the `doc` folder as `uml_overview.txt`. Feel free move it of course.

The result is as follows. I'm curious to hear your feedback.

![uml_overview 06-commit-865991f](https://user-images.githubusercontent.com/3320415/111039229-3e62f700-842d-11eb-8858-3a7afea93a71.png)

titusjan commented:
Oops. I already noted I made a mistake. The `ImageView` class contains a `pq.GraphicsView` attribute, not a `QGraphicsView` attribute. I fixed it. I changed one line and half the classes have a different location lol.

![uml_overview 07-commit-099fc23](https://user-images.githubusercontent.com/3320415/111041777-407f8280-843a-11eb-8364-49e4343f4733.png)

jotaf98 commented:
This looks great, and is really useful! Could I suggest referencing the image as a page in the docs, with a one-liner saying what the colors, diamonds and arrows represent?

j9ac9k commented:
@titusjan Sorry I haven't followed up, we're still crafting our proposal for the grant we're applying for, and I'm trying to wrap up my PR from forever ago (1318); ... I want to experiment and see what this looks like with more of a hierarchy. 

The other thing I'm going to want to see if viable is making a dark mode variant; I embed pyqtgraph docs within [dash](https://kapeli.com/dash) and it looks like this presently:

![image](https://user-images.githubusercontent.com/646398/111653950-87b49d00-87c5-11eb-9f47-f91a0f1d136d.png)

I think having one image that works for both light and dark mode won't work, and I'm not sure how "easy" it would be to have multiple images that the docs can pick from for this sort of thing either...

titusjan commented:
> Sorry I haven't followed up

No worries @j9ac9k. I understand that maintaining PyQtGraph takes a lot of your time. I'm already impressed by the amount of time you spend on it.

>  I want to experiment and see what this looks like with more of a hierarchy.

What do you mean by this? If you mean using PlantUML name spaces then I would say don't do it :-)
I tried this in the beginning and I think this makes it very cluttered and unclear (see below). This is because the Qt classes and PyQtGraph classes are at many different levels in the dependency graph. It's **not** case that the Qt classes are on top and the PyQtGraph classes are at the bottom.

![uml_overview 01-packages](https://user-images.githubusercontent.com/3320415/111661043-f8cc7280-880e-11eb-9fb7-6c8bb92d08c7.png)

If you make a dark mode variant, please use different colors for the inheritance lines and the aggregation lines. This greatly improves the readability IMO.

j9ac9k commented:
> No worries @j9ac9k. I understand that maintaining PyQtGraph takes a lot of your time. I'm already impressed by the amount of time you spend on it.

The contributors make it worth the effort 👍🏻 

Regarding the hierarchy, I definitely would want the Qt and PyQtGraph objects together (to better highlight their relationship to one another).  I meant that the inheritance is demonstrated via vertical spacing.

I just sketched something out on a tablet, hopefully it makes a bit more sense.

![Quick sheets - page 4](https://user-images.githubusercontent.com/646398/111666519-07943480-87d1-11eb-8420-3368802dd84f.png)

Idea here being that the the "parent" objects are at the top, and the inheritance flows in a vertical space.  From the looks of your most recent screenshot, it looks like you largely have that (you just have the Qt and PyQtGraph classes also separated).

j9ac9k commented:
re-looking at my picture, I want to point out that `GraphicsItem`, `QGraphicsView` and `QGraphicsScene` should likely be laid out horizontally at the "top" of the plot, not vertically as I have done (did this due to my tablet being in portrait mode, ...probably should have rotated it first :|

titusjan commented:
> Idea here being that the the "parent" objects are at the top, and the inheritance flows in a vertical space. From the looks of your most recent screenshot, it looks like you largely have that 

...

> I want to point out that GraphicsItem, QGraphicsView and QGraphicsScene should likely be laid out horizontally at the "top" of the plot

In the first two screen shots parent classes are positioned above their children in **all** cases. Also the GraphicsItem, QGraphicsView and QGraphicsScene classes are nearly at the top. It's just that their parents and the ImageView class (which contains a QGraphicsScene) are located above them.

I'm curious to see your diagrams when they are finished.

j9ac9k commented:
First off, I just want to reiterate, this diagram is amazing, and very much needed.  I want this merged into the documentation, thanks for this PR.

> In the first two screen shots parent classes are positioned above their children in all cases. Also the GraphicsItem, QGraphicsView and QGraphicsScene classes are nearly at the top. It's just that their parents and the ImageView class (which contains a QGraphicsScene) are located above them.

You do have that, I got a little crossed up by the curved/angled/overlapping lines, my apologies.  I had a mental model of lines that were straight at right angles (horizontal or vertical); might not be best in this case though.  Also not sure how feasible it would be to not have overlapping darker lines for inheritance (for the lighter colored lines will have to overlap, but given what they represent, I think that's ok).

I'm talking myself in and out of the benefits of having `QObject` listed, on one hand, (just about) _everything_ except GraphicsItem points to `QObject`, and `QObject` does have methods/properties which could be of use; probably should keep it.

j9ac9k commented:
and that diagram reminded me...

![image](https://user-images.githubusercontent.com/646398/111682819-09b2bf00-87e2-11eb-86f5-fca76f2e4e16.png)

We should `probably` modify `PlotDataItem` to use `GraphicsLayout` for a layout instead of `QGraphicsGridLayout`.... but that's a whole other issue!

lyl1836 commented:
@j9ac9k ：

The picture of your help file is very good!

Where can I get it?

[https://user-images.githubusercontent.com/646398/111653950-87b49d00-87c5-11eb-9f47-f91a0f1d136d.png](url)

j9ac9k commented:
Hi @titusjan I'm playing with this in the PlantUML website; is there a UML example/tutorial/guide you like?  I want to experiment with this some.

titusjan commented:
@j9ac9k I only read the Class Diagram chapter of the [reference guide](http://plantuml.com/guide). It contains a lot of examples so it worked for me.

j9ac9k commented:
ok, I'm messing with it now, PlantUML is _really_ punishing if you try and force the situation with a layout isn't it...

The following UML code generates this:

```plantuml
@startuml
' Made in PlantUml. You can try it out here: http://plantuml.com/plantuml
' Based on commit eb7a60fcf83cd4e7a41ae5955e57935e39928fbd


hide empty members
hide circles
hide stereotypes
top to bottom direction

skinparam class {
    BorderColor Black
    ArrowColor Black

    ' Using stereotypes to define the background colors
    BackgroundColor<<qt>> #e5f2da
    BackgroundColor<<pg>> #e5e5f4
}
skinparam shadowing false


'----------- Qt package ----------`


class QGraphicsGridLayout <<qt>>

class QGraphicsItem <<qt>>

class QGraphicsLayout <<qt>>

class QGraphicsLayoutItem <<qt>>

class QGraphicsObject <<qt>>

class QGraphicsScene <<qt>> { 
    items 
}

class QGraphicsWidget <<qt>>

class QGraphicsView <<qt>> {
    scene
}

class QObject <<qt>>

class QPaintDevice <<qt>>

class QWidget <<qt>>


'----------- PyQtGraph package ----------`


class GraphicsItem <<pg>>

class GraphicsLayout <<pg>> {
    layout
}

class GraphicsLayoutWidget <<pg>> {
    graphicsLayout
}

class GraphicsObject <<pg>>

class GraphicsView <<pg>>

class GraphicsWidget <<pg>> 

class ImageItem <<pg>>

class ImageView <<pg>> {
    graphicsView
    imageItem
    scene
    view
}

class PlotCurveItem <<pg>>

class PlotDataItem <<pg>> {
    curve
    scatter
}

class PlotItem <<pg>> {
    layout
    vb
}

class PlotWidget <<pg>> {
    plotItem
}

class ScatterPlotItem <<pg>>

class ViewBox <<pg>>


'---------- Inheritance within Qt ----------'
QGraphicsItem <-- QGraphicsObject
QObject <-- QGraphicsObject
'QGraphicsItem -[hidden]> QObject
QGraphicsObject <-- QGraphicsWidget
QGraphicsLayoutItem <-- QGraphicsWidget
QGraphicsLayoutItem <-- QGraphicsLayout
QGraphicsLayout <-- QGraphicsGridLayout
QPaintDevice <-- QWidget
QObject <-- QWidget
QObject <-- QGraphicsScene
QWidget <-- QGraphicsView


'---------- Inheritance from Qt to PyQtGraph ----------'
QGraphicsWidget <-- GraphicsWidget
QGraphicsObject <-- GraphicsObject
QGraphicsView <-- GraphicsView
QWidget <---- ImageView


'---------- Inheritance  within PyQtGraph ----------'
GraphicsItem <-- GraphicsObject
GraphicsItem <-- GraphicsWidget
GraphicsWidget <-- GraphicsLayout
GraphicsWidget <-- PlotItem
GraphicsWidget <-- ViewBox
GraphicsObject <--- ScatterPlotItem
GraphicsObject <--- PlotCurveItem
GraphicsObject <--- ImageItem
GraphicsObject <-- PlotDataItem
GraphicsView <-- PlotWidget
GraphicsView <-- GraphicsLayoutWidget


'---------- Aggregation ----------'

' Shorter arrow so at same height in the diagram
QGraphicsScene::items o-[norank]- QGraphicsItem                      #b8b8b8
QGraphicsView::scene o-[norank]- QGraphicsScene                      #b8b8b8

' Longer (regular) arrows
PlotWidget::plotItem o-[norank]- PlotItem                           #b8b8b8
GraphicsLayoutWidget::graphicsLayout o-[norank]- GraphicsLayout     #b8b8b8
PlotDataItem::curve o-[norank]- PlotCurveItem                       #b8b8b8
PlotDataItem::scatter o-[norank]- ScatterPlotItem                   #b8b8b8
PlotItem::vb o-[norank]- ViewBox                                    #b8b8b8
PlotItem::layout o-[norank]- QGraphicsGridLayout                    #b8b8b8
GraphicsLayout::layout o-[norank]- QGraphicsGridLayout              #b8b8b8
ImageView::graphicsView o-[norank]- GraphicsView                    #b8b8b8
ImageView::imageItem o-[norank]- ImageItem                          #b8b8b8
ImageView::scene o-[norank]- QGraphicsScene                         #b8b8b8
ImageView::view o-[norank]- ViewBox                                 #b8b8b8
@enduml
```

![image](https://user-images.githubusercontent.com/646398/112729413-31dca500-8ee9-11eb-9745-08c5a789ceb0.png)

titusjan commented:
> ok, I'm messing with it now, PlantUML is really punishing if you try and force the situation with a layout isn't it...

Yes indeed :-). And small changes in the connections lead to big changes in the layout. I more or less stopped trying to fight it.

j9ac9k commented:
Yeah my strategy now is larger to specify which arrows should have no rank,
and which arrows (if any) should be longer than others.   This methodology
does "feel" right because it lets me declare what is important and how
important and how distant relationships are.

On Sat, Mar 27, 2021 at 11:13 Pepijn Kenter ***@***.***>
wrote:

> ok, I'm messing with it now, PlantUML is really punishing if you try and
> force the situation with a layout isn't it...
>
> Yes indeed :-). And small changes in the connections lead to big changes
> in the layout. I more or less stopped trying to fight it.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pyqtgraph/pyqtgraph/pull/1631#issuecomment-808776776>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE5Z7XPFOXCEQVIB5MN5G3TFYN3RANCNFSM4ZEBNOLQ>
> .
>

j9ac9k commented:
Ok, I fixed some issues involving `GraphicsScene` and removed one component of `ImageView` that I didn't think was needed to reference

![image](https://user-images.githubusercontent.com/646398/112732829-9273de00-8ef9-11eb-9a37-f34eaf9a3116.png)


```plantuml
@startuml
' Made in PlantUml. You can try it out here: http://plantuml.com/plantuml
' Based on commit eb7a60fcf83cd4e7a41ae5955e57935e39928fbd


hide empty members
hide circles
hide stereotypes
top to bottom direction

skinparam class {
    BorderColor Black
    ArrowColor Black

    ' Using stereotypes to define the background colors
    BackgroundColor<<qt>> #e5f2da
    BackgroundColor<<pg>> #e5e5f4
}
skinparam shadowing false


'----------- Qt package ----------`


class QGraphicsGridLayout <<qt>>

class QGraphicsItem <<qt>>

class QGraphicsLayout <<qt>>

class QGraphicsLayoutItem <<qt>>

class QGraphicsObject <<qt>>

class QGraphicsScene <<qt>> { 
    items 
}

class QGraphicsWidget <<qt>>

class QGraphicsView <<qt>> {
    scene
}

class QObject <<qt>>

class QPaintDevice <<qt>>

class QWidget <<qt>>


'----------- PyQtGraph package ----------`


class GraphicsItem <<pg>> {
    scene
}

class GraphicsLayout <<pg>> {
    layout
}

class GraphicsScene <<pg>>

class GraphicsLayoutWidget <<pg>> {
    graphicsLayout
}

class GraphicsObject <<pg>>

class GraphicsView <<pg>> {
    scene
}

class GraphicsWidget <<pg>> 

class ImageItem <<pg>>

class ImageView <<pg>> {
    graphicsView
    imageItem
    view
}

class PlotCurveItem <<pg>>

class PlotDataItem <<pg>> {
    curve
    scatter
}

class PlotItem <<pg>> {
    layout
    vb
}

class PlotWidget <<pg>> {
    plotItem
}

class ScatterPlotItem <<pg>>

class ViewBox <<pg>>


'---------- Inheritance within Qt ----------'
QGraphicsItem <-- QGraphicsObject
QObject <-- QGraphicsObject
'QGraphicsItem -[hidden]> QObject
QGraphicsObject <-- QGraphicsWidget
QGraphicsLayoutItem <-- QGraphicsWidget
QGraphicsLayoutItem <-- QGraphicsLayout
QGraphicsLayout <-- QGraphicsGridLayout
QPaintDevice <-- QWidget
QObject <-- QWidget
QObject <-- QGraphicsScene
QWidget <-- QGraphicsView


'---------- Inheritance from Qt to PyQtGraph ----------'
QGraphicsWidget <-- GraphicsWidget
QGraphicsObject <-- GraphicsObject
QGraphicsView <-- GraphicsView
QWidget <--- ImageView
QGraphicsScene <-- GraphicsScene


'---------- Inheritance  within PyQtGraph ----------'
GraphicsItem <-- GraphicsObject
GraphicsItem <-- GraphicsWidget
GraphicsWidget <-- GraphicsLayout
GraphicsWidget <-- PlotItem
GraphicsWidget <-- ViewBox
GraphicsObject <-- ScatterPlotItem
GraphicsObject <-- PlotCurveItem
GraphicsObject <-- ImageItem
GraphicsObject <-- PlotDataItem
GraphicsView <-- PlotWidget
GraphicsView <-- GraphicsLayoutWidget


'---------- Aggregation ----------'

' Shorter arrow so at same height in the diagram
QGraphicsScene::items o-[norank]- QGraphicsItem                      #b8b8b8
QGraphicsView::scene o-[norank]- QGraphicsScene                      #b8b8b8
GraphicsView::scene o-[norank]- GraphicsScene                      #b8b8b8


' Longer (regular) arrows
PlotWidget::plotItem o-[norank]- PlotItem                           #b8b8b8
GraphicsItem::scene o-[norank]- GraphicsScene                       #b8b8b8
GraphicsLayoutWidget::graphicsLayout o-[norank]- GraphicsLayout     #b8b8b8
PlotDataItem::curve o-[norank]- PlotCurveItem                       #b8b8b8
PlotDataItem::scatter o-[norank]- ScatterPlotItem                   #b8b8b8
PlotItem::vb o-[norank]- ViewBox                                  #b8b8b8
PlotItem::layout o-[norank]- QGraphicsGridLayout                    #b8b8b8
GraphicsLayout::layout o-[norank]- QGraphicsGridLayout              #b8b8b8
ImageView::graphicsView o-[norank]- GraphicsView                    #b8b8b8
ImageView::imageItem o-[norank]- ImageItem                          #b8b8b8
ImageView::view o-[norank]- ViewBox                                 #b8b8b8
@enduml
```

j9ac9k commented:
tool-tips and hyperlinks work when exported as SVG! ...hopefully that means they'll stay working when embedded with sphinx.  I'll have to test this with a branch on the main repository.

[FYI, this is what I currently have][(http://www.plantuml.com/plantuml/png/bLPVJoCt47_tfnZf7j8-1Dg355YTq1rtqXM9ga4ShIg4LCSxsNNPjHTx0eqGtxthECjcDYOzdAVC_EOtVsOywu-UfAD5NOaH_2Pp16rWMab3rtMr1t_P1IXfWDmID85T49JeC8EIgCdszvk0NBH8PUlu9pATIOyvM0EjlDO4E3kIl_mqL_FZ0vKVuf4y_5dYv6GomSdHoS443qvEtX_FPxaGejGv0jODBQ76UeREHv7IJbMu_kC97LfQDgs0R0DaOMQ9R0svTgX8MoE4lzEcaKxMe2hfFJm9QC-PTJcwpxQo3iuggUvMqa_EsSTDuKewWcklJR7fBhZASQuD0fK8inPREBimEQXWxQEF5-cAyiE7UpezXNSucR_FPHhH517HOWx5yqRilfIvVGnHp6NbKGWn6lS7BWcQbaiM2BtqxnOLSxxywcHJQkM_EfrVo6NeO0nd2t5EMB-cuvPfxIxxtsV_e7hL_agXmRKIdc1L8DsoUH3FM-0_TLxWgrH_Q7pic6vkmfJwTanpg_RkQK_R_NiQJ_RlYpNweKNlbLHNJz-ihRxfvXdWrxRBxaARk-VRMtZXJoSmbThG5tpG2eUgGPYyQTFb9QqYtjaxtfWu8N66VIXONnlU1mwjLZA6PJKFq3HFp82pASsPM6yv5Ul7_yVElNNQyreMk94_aszp5njUunHrzgj_eTszvsbbwVF2FN3-NlT5ajmklGeMwqGa4JfEo0tw1aJ_CmvELRXPa_J8g-XdmDvfGwfdzjzEoWODpas9JfCq2k5HKwbDs1SzO2G6jx-r6TnPqOr-IZVYvkERKkSvcjlJxiBq_8o5zLmaziZRS5m-n75ClmS5lxu151tnb5CoTeb4X31L7C4TtPWxMuTUaEsNGRen6-HCbAuj5maM3WC4mMRau_vE2PvVPzPBTkJLZLawAJOjYN2PFf5niYQizmb4Tt4IkkxsY4Gb1zSk0U4R90M8wokjP3j6RFMftm_RkjIU7lJaKr4uB2HfQrW7n0YkIki87SZmv05lGH9uMIEKg8kIG9lLooRNid2o7Wn3biKFintV6EkakRiT3ruGoVDkTXn-B9Kiyu6HKJ5Vkwb2AXVM5EZW1uV5efBknvYJ5ttnignRgCpFj1EcJoha07zxo9mhrRWiKveHSX2ptnoPrYwC7ZFbdxJqINEjFsQCRVZXsSqLUHvcZABxEkq-IQRutMHiYIN-_TL-AoFdUbcEMRRvo4WrBYeI9y7ryZvXH06GQk5khl2wUMFnETT7DFcYhlu3) I'm debating if I want to add more objects (axis items come to mind).

@titusjan how "complete" should we try and get this diagram do you think?

j9ac9k commented:
@titusjan would you be ok with me pushing commits on this branch, attempting to embed this PlantUML code within the docs?

titusjan commented:
@j9ac9k. Sorry for not reacting sooner. I'm busy at work so it's hard to find the free time. I'm still looking in to this. Would you mind waiting a little bit longer? I will try to give a reply this evening.

j9ac9k commented:
No problem, due to other events us maintainers have been involved with, documentation is front and center on our mind

I offered to help because I think it might be a bit tricky to merge the plantuml code into readthedocs.  Let me know if I can help.

titusjan commented:
Again apologies for staying offline for such a long time. Life gets in the way sometimes.

@j9ac9k I like your latest version. It is less cluttered and is certainly an improvement regarding the inheritance relations. However, IMO this is a bit at the expense of the aggregation relationships. I think the diagram becomes much easier to read if the dependencies of a class are located below that class, or at the same level and to the right. Now all the black arrows of the inheritance relations go down, but a lot of the grey aggregation lines go back up again. 

It may not always be possible to achieve this goal but I'd like to try. In particular I think that the `QGraphicsItem` should not be located above the `QGraphicsScene` but below it or to the right. The same for the `QGraphicsScene` and the `QGraphicsView`. I tried to place them in a 3 by 2 grid together with their PyQtGraph equivalents, but when I removed the `norank` constraints the resulting diagram became a total mess. 

As you also noticed PlantUML gives us little control over the layout. I hope you don't mind but I therefore recreated the diagram in [yEd](https://www.yworks.com/products/yed). This is a graph editor that is easy to use. In contrast to Dia it saves the graphs as text (XML), so that they can be included in git.

I used your latest version as the basis. In yEd it looks as follows (commit [0c39928](https://github.com/pyqtgraph/pyqtgraph/pull/1631/commits/0c39928589014907529c2d7edf51e31d683c755c)). 

![overview_uml 01_manual](https://user-images.githubusercontent.com/3320415/118114878-7617fc00-b3e8-11eb-90ba-9e75b9cb23cf.png)

Next, I flipped it horizontally so that `QGraphicsView` is to the left of `QGraphicsScene` ([04a73a4](04a73a43e316b636c603602d62e2ef15d3704fad).) 

![overview_uml 02_flipped](https://user-images.githubusercontent.com/3320415/118115313-13733000-b3e9-11eb-9d08-55cf5ecc17ea.png)

I then moved some classes up or down so that `(Q)GraphicsView`, `(Q)GraphicsScene` and `(Q)GraphicsItem` are all at the same level. After some further tweaking this is what I've got now ([9ee3d8d](9ee3d8d985087f39b461d3205dca5846e9585ab0)). Let me know what you think.

![overview_uml 03_latest](https://user-images.githubusercontent.com/3320415/118115915-da878b00-b3e9-11eb-8ecf-cf3219dd8bdb.png)

Finally, I checked again what attributes the `(Q)GraphicsView`, `(Q)GraphicsScene` and `(Q)GraphicsItem` classes have, and fixed them accordingly. (BTW, thanks for spotting that the `GraphicsScene` class was missing in my original version).

Summarized:

- QGraphicsView: has `scene()` and `items()` methods (not attributes, so I changed this in the diagram.)
- QGraphicsScene: has `views()` and `items()` methods.
- QGraphicsItem: has a `scene()` method.
- GraphicsView: has a `sceneObj` attribute. This seems to be for internal use only (to keep a reference), so we could decide to leave it out.
- GraphicsScene: overrides `items()` but only to a fix bug in Qt. It still returns a list of `QGraphicsItems`. Therefore I didn't add a method in the diagram.
- GraphicsItem: doesn't define or override the `scene()` method so I removed it from the diagram. It does, however, use it in `getViewWidget` and a few other methods. This probably only works because nobody uses a `GraphicsItem` directly.

I'm curious about your feedback.

NilsNemitz commented:
Dear @titusjan , thank you for the work you put into this.
I do think that your latest version is the clearest so far, although I haven't been following the details of the discussion very closely.

I'd like to point out a minor thing in case you have a good idea how to include it in the diagram:
There is a clear hierarchy between PlotItem and the classes derived from GraphicsObject: 
ImageItem, PlotDataItem, ScatterPlotItem, PlotCurveItem etc. are all things that will be held and managed by a PlotItem. That relationship unfortunately is not very clear from the naming. and I think new users might benefit from additional ways to communicate this.

So even if this is not really the main point of the class diagram, it might be helpful to move the X-Items to appear below PlotItem. Maybe pull GraphicsObject and its derived classes to the bottom of the current diagram?
A single aggregation arrow from PlotItem to GraphicsObject might help communicate the relation, even though I am not sure if that is technically correct.

j9ac9k commented:
Hi @titusjan thanks for following up on this.

I'm a firm believer of not letting perfection be the enemy of good, and that something is far better than nothing in this area; and so I'll be supportive of a merge even if I disagree with some of the design elements.  This is a long winded way of me saying, for design stuff, please do not feel obligated to implement my suggested changes if you do not think it is a good idea.

On the subject of format, my concern with whatever format we use here is in the form of integration into the docs.  PlantUML is somewhat nice in that regard as readthedocs has support for compiling PlantUML code into SVGs, which support hyper-linking/tool-tips (which I thought would be nice additions); since yEd supports XML here, that functionality is likely available there as well.  I don't see why we can't include both though (yEd for the large diagram that you have in your PR, and plantUML for smaller diagrams to show hierarchy of individual objects,...another way of me saying, while I prefer PlantUML, it's not a "must have", provided alternative solutions work w/ readthedocs.

To the styling of this plot, I like the layout.  I'm in agreement with @NilsNemitz about trying to better show the relationship between PlotItem and some of the children of GraphicsObject.  I'm not strongly opinionated on how that should be done.

Next question I suppose is the integration into the documentation, since we have readthedocs run on PRs, we should be able to see what the result looks like there.

Thanks again for tackling this @titusjan!

titusjan commented:
Hi @j9ac9k. Glad you like the layout.

It is indeed possible in yEd to export to SVG and to include tooltips and hyperlinks. So I added hyperlinks to the documentation and added the first one or two relevant sentences as tooltips. Please let me know if you want to see something else in the tooltips.

I added the exported SVG file to the git repo so that it can be processed by read-the-docs. 

> I don't see why we can't include both though (yEd for the large diagram that you have in your PR, and plantUML for smaller diagrams to show hierarchy of individual objects

I agree.

@NilsNemitz, I will think on how to best show the relationship between PlotItem and GraphicsObject.

titusjan commented:
Hi @NilsNemitz and @j9ac9k 

Looking at the source code I found that `PlotItem` has an `items` attribute that contains a list of `QGraphicsItem` references. The `PlotItem.addItem()` methods adds an item to this list, and also calls `ViewBox.addItem()`, which adds the item to the scene. I added these to the diagram and put this information in the tooltips (which are a great idea BTW). 

So the `addItem` methods accept any `QGraphicsItem` object and therefore it's probably technically the most correct to let the aggregation line go from `PlotItem.items` to  the`QGraphicsItem` class. This is done in the diagram below. However, if you prefer the line to go to the `GraphicsObect` class let me know and I'll change it.

![overview_uml](https://user-images.githubusercontent.com/3320415/118394169-47e22880-b643-11eb-8c58-9e45cfa04276.png)

NilsNemitz commented:
@titusjan , thank you for looking at that!

But I am afraid that pointing out that basically anything can be added to a PlotItem, potentially in an infinite recursion, may be more confusing than enlightening for the user. 

I was hoping to see the following point communicated:
`PlotItem` typically holds and manages one or more of `ImageItem`, `PlotDataItem`, `ScatterPlotItem` or `PlotCurveItem`.

If there isn't a good way to illustrate this in the diagram, that is fine... We'll find some other place to do it.
So I would prefer the previous version, without the `addItem` reference; at least unless you, @j9ac9k or someone else has a genius idea. Sorry for the trouble!

titusjan commented:
Hi @NilsNemitz .

No problem. I can see why you prefer a line to `GraphicsObject` as it is typically used that way. Like i said it's no problem to change it. What do you think of the diagram below?

![overview_uml](https://user-images.githubusercontent.com/3320415/118396354-7b2ab480-b64f-11eb-9a90-74593c17b1b5.png)

NilsNemitz commented:
@titusjan : I like it... Any chance of pulling GraphicsObject and its derivatives down until they sit below PlotItem?

titusjan commented:
> Any chance of pulling GraphicsObject and its derivatives down until they sit below PlotItem?

This makes the diagram larger and perhaps less aesthetically pleasing. However does make the dependencies flow downwards, which I advocated earlier. See the update below. What do you guys think?

![overview_uml](https://user-images.githubusercontent.com/3320415/118398149-98638100-b657-11eb-9a03-c4ef4516179b.png)

NilsNemitz commented:
Well, I kind of like it :)
My feeling is that it might look better, or at least have less crossed lines if
- the full column from QGraphicsLayoutItem to GraphicsLayout moved two steps to the left.
- GraphicsItem and ViewBox moved two steps to the right
- GraphicsObject moved all the way to the right.

But maybe we get some other opinions before I make you mess with the plot even more :)

titusjan commented:
@NilsNemitz I'd like to keep `GraphicsItem` directly below `QGraphicsItem`. The trigger for me to start again in yEd was to have full control so that `(Q)GraphicsView`, `(Q)GraphicsScene`, and `(Q)GraphicsItem` could be placed in a 3 by 2 grid.

> But maybe we get some other opinions before I make you mess with the plot even more :)

Yes, let's hear the opinion of @j9ac9k about the last two versions.

j9ac9k commented:
Generally speaking, I'm going to be partial to compact layouts with minimal crossed thick/bolt lines.  The reason I am partial to the compactness is that this will need to be displayed on a display, and I would want to minimize the likelihood of the user having to scroll; especially as they may have some code-alongside the display.

j9ac9k commented:
Hi @titusjan 

I'd like to get this PR merged before much longer; I'm likely going to spend some time trying to integrate the construction of the chart that you did into our CI pipeline (or see if it's feasible at least).

Sorry we let this get stale!

titusjan commented:
> Sorry we let this get stale!

No problem. I didn't follow up either.

I notice that we never made a final decision between the last two options. I see that you prefer the compact layout. I assume you mean [this one](https://github.com/pyqtgraph/pyqtgraph/pull/1631#issuecomment-841808152). That one has my preference as well, so let's go with that.

I'm curious to the final result.

j9ac9k commented:
I do prefer the one you linked, but my feelings between the two are close enough, that I would let the author have the discretion of which one they want to use.

I loaded the graph inside yEd; everything looks good there 👍🏻 

My main gripe/sticking point is that I'm not seeing an obvious way to generate the SVG as part of the docs build process, so we can have it in the CI.  _Ideally_ we would be able to have just the code, and not have to worry about the SVG being committed to the repo.  I'll try and dig around and see if there is a way I can build/export the SVG from that file via an automated process outside of the desktop app.

I should also note, that while that is my main sticking point, I no longer feel that not being able to export the image in the CI pipeline is a deal breaker.  This chart has a sufficient utility that I would be willing to handle some inconvenience.

What should likely also happen is a reference to this diagram, and how to update it should be made to `CONTRIBUTING.md`

j9ac9k commented:
Ok, we can't use this as part of an automated build process:

From their [license agreement](https://www.yworks.com/resources/yed/license.html):

> The Licensee is granted a non-exclusive and non-transferable right to install one copy of the Software and use it as an application. **The Software may not be used as part of an automated process.** The Licensee may not reverse engineer, disassemble, decompile, or unjar the Software, or otherwise attempt to derive the source code of the Software.

j9ac9k commented:
I started tinkering with mermaid; and I'll say right off the bat, the output doesn't look anywhere near as good as what you have created using yEd; that said, I think we can do away w/ the light arrows and just specify the types in the entries.

There is something nice about the code being just that simple tho... this size diagram might just be too big for mermaid to render nicely.

```mermaid
classDiagram
    QPaintDevice <|-- QWidget
    QObject <|-- QWidget
    QObject <|-- QGraphicsScene
    QObject <|-- QGraphicsObject

    QWidget <|-- ImageView
    QWidget <|-- QGraphicsView

    QGraphicsItem <|-- QGraphicsObject

    QGraphicsView <|-- GraphicsView

    QGraphicsScene <|-- GraphicsScene

    GraphicsView <|-- GraphicsLayoutWidget

    GraphicsView <|-- PlotWidget

    GraphicsItem <|-- GraphicsObject
    GraphicsItem <|-- GraphicsWidget
    QGraphicsObject <|-- GraphicsObject
    QGraphicsObject <|-- QGraphicsWidget
    QGraphicsLayoutItem <|-- QGraphicsWidget
    QGraphicsLayoutItem <|-- QGraphicsLayout

    GraphicsObject <|-- ImageItem
    GraphicsObject <|-- ScatterPlotItem
    GraphicsObject <|-- PlotDataItem
    GraphicsObject <|-- PlotCurveItem

    QGraphicsWidget <|-- GraphicsWidget
    QGraphicsGridLayout <|-- QGraphicsLayout

    GraphicsWidget <|-- ViewBox
    GraphicsWidget <|-- PlotItem
    GraphicsWidget <|-- GraphicsLayout

    class GraphicsLayout{
        layout: QGraphicsGridLayout
    }

    class PlotItem{
        items: List[GraphicsItem]
        layout: QGraphicsGridLayout
        vb: ViewBox
        addItem(GraphicsItem)
    }

    class ViewBox{
        addItem(GraphicsItem)
    }

    class PlotDataItem{
        curve: PlotCurveItem
        scatter: PlotScatterItem
    }

    class PlotWidget{
        plotItem: PlotItem
    }


    class QGraphicsView{
        scene()
    }
    class ImageView{
        graphicsView: GraphicsView
        imageItem: ImageItem
        view: ViewBox
    }

    class GraphicsLayoutWidget{
        graphicsLayout: GraphicsLayout
    }
```

titusjan commented:
The YEd license says the software can't be used as part as of an automated build process but I understand this as that it can't be executed automatically. I think it is perfectly legal to just include the generated SVG in the repo. 

>  Ideally we would be able to have just the code, and not have to worry about the SVG being committed to the repo. 

You will always have to make a selection of classes and tweak the generated output. Even with PlantUml there was a large config file that needs to be maintained. IMO there's no difference between maintaining the config file or directly maintaining a diagram. Or am I missing something?

>  I think we can do away w/ the light arrows and just specify the types in the entries.

We could also do both :-)

j9ac9k commented:
> You will always have to make a selection of classes and tweak the generated output. Even with PlantUml there was a large config file that needs to be maintained. IMO there's no difference between maintaining the config file or directly maintaining a diagram. Or am I missing something?

The issue is that we can't "generate" the entirety of the documentation with a `make` command as we (more or less) currently do.  There are [images](https://github.com/pyqtgraph/pyqtgraph/tree/master/doc/source/images) like that in the docs right now, so it's not like we're introducing something new with this PR (although some of those images could easily be migrated to PlantUML/Mermaid/GraphiViz generated images, but that's for another topic).

As you correctly point out, this is likely not as big of a deal as I'm making it out to be :) 

Again, don't think this is a deal breaker, I'm just trying to make sure there isn't a platform that would work better for us.  Since this PR was opened, mermaid has gotten much more support (renders in github markdow for example).  

Assuming we stay w/ yEd (which I have yet to be convinced of another diagram platform that is more suitable); a blurb to `CONTRIBUTING.md` should be added about this image/file and how it was created and the application needed to edit it.   Lastly, the image should go live here: https://github.com/pyqtgraph/pyqtgraph/tree/master/doc/source/images

I'm open to suggestions on where in the docs the image should live.

j9ac9k commented:
Hi @titusjan if you have the opportunity, could you update the graph with the [condensed version](https://github.com/pyqtgraph/pyqtgraph/pull/1631#issuecomment-841808152); from that point, I can figure out where to integrate into the docs and such.  

I think after this is merged, we will release 0.13.0 👍🏻

titusjan commented:
Sorry for not responding earlier. I was on a short vacation and didn't read my email.

I reverted back to the [condensed version](https://github.com/pyqtgraph/pyqtgraph/pull/1631#issuecomment-841808152) and moved it to the `doc/source/images` directory.

> I'm open to suggestions on where in the docs the image should live.

Perhaps a separate page in the API reference would be the best. It can contain a small paragraph explaining the image. I think there should be a hyperlink to the diagram somehwere in the [Plotting in pyqtgraph](https://pyqtgraph.readthedocs.io/en/latest/plotting.html) as well.

ixjlyons commented:
I view the "Plotting in pyqtgraph" page as introductory/tutorial material, and putting the diagram there might be a little overwhelming to the target reader. I like the idea of giving it a separate page, probably under API reference, and linking to it.

We're also lacking [internals docs](https://pyqtgraph.readthedocs.io/en/latest/internals.html) and this would be a great reference there too, once there's some expository content. I've always been tempted to put a placeholder there.

![download](https://user-images.githubusercontent.com/943602/191804485-c1bf7889-ac7c-411c-bd0b-2a33c7749184.jpg)

j9ac9k commented:
> Perhaps a separate page in the API reference would be the best. It can contain a small paragraph explaining the image. I think there should be a hyperlink to the diagram somehwere in the [Plotting in pyqtgraph](https://pyqtgraph.readthedocs.io/en/latest/plotting.html) as well.

Would you like me to create this and add to this branch or would you want to draft something up?

titusjan commented:
> Would you like me to create this and add to this branch or would you want to draft something up?

I made a draft version. Let me know what you think or feel free to make modifications.

j9ac9k commented:
> > Would you like me to create this and add to this branch or would you want to draft something up?
> 
> I made a draft version. Let me know what you think or feel free to make modifications.

Just read it over, the contents look good to me.  The only thing preventing the merge is the merge-conflict in `apireference.rst`.

titusjan commented:
I merged the upstream master branch into my issue-848 branch and resolved the conflict.

titusjan commented:
Yes it was a reference to the "Deprecated Window Classes" page, which seems to have been removed from the documentation. 

I removed the line.

j9ac9k commented:
RTD passes, when CI is green, I'm going to merge.

Thanks, @titusjan for your work on this, and your extraordinary patience with us.   Should you submit another PR, I'll do the best I can to make sure it doesn't take 1.5 years before merging!

titusjan commented:
Thanks for your support and congrats on the upcoming release. Your work is greatly appreciated @j9ac9k

j9ac9k commented:
@titusjan do you have a twitter handle I can tag you with relating to this PR? I'd like to post to twitter and I'd like to give you credit.

titusjan commented:
No I don't have twitter. The titusjan account on twitter was already taken by someone else unfortunately.

Thanks for wanting to giving me credit though.

j9ac9k commented:
Hi @titusjan 

Slight follow-up here, do you know if yEd allows for storing multiple color palettes in a file, allowing one to quickly toggle between the two?  The motivation here is to generate a dark mode friendly svg and a light mode friendly one... without having two separate .graphml files.

titusjan commented:
I couldn't find any functionality like that, unfortunately.

j9ac9k commented:
> I couldn't find any functionality like that, unfortunately.

Thanks for checking!
